integration/gotest_test.go

```diff
@@ -0,0 +1,162 @@
+		"github.com/u-root/u-root/cmds/modprobe",
+		"github.com/u-root/u-root/cmds/mount",
+		"github.com/u-root/u-root/cmds/msr",
+		"github.com/u-root/u-root/cmds/netcat",
+		"github.com/u-root/u-root/cmds/pci",
+		"github.com/u-root/u-root/cmds/pe",
+		"github.com/u-root/u-root/cmds/ping",
+		"github.com/u-root/u-root/cmds/pwd",
+		"github.com/u-root/u-root/cmds/pxeboot",
+		"github.com/u-root/u-root/cmds/rmmod",
+		"github.com/u-root/u-root/cmds/rsdp",
+		"github.com/u-root/u-root/cmds/sos",
+		"github.com/u-root/u-root/cmds/srvfiles",
+		"github.com/u-root/u-root/cmds/sshd",
+		"github.com/u-root/u-root/cmds/stty",
+		"github.com/u-root/u-root/cmds/switch_root",
+		"github.com/u-root/u-root/cmds/sync",
+		"github.com/u-root/u-root/cmds/tail",
+		"github.com/u-root/u-root/cmds/tcz",
+		"github.com/u-root/u-root/cmds/tee",
+		"github.com/u-root/u-root/cmds/timesos",
+		"github.com/u-root/u-root/cmds/umount",
+		"github.com/u-root/u-root/cmds/uname",
+		"github.com/u-root/u-root/cmds/unshare",
+		"github.com/u-root/u-root/cmds/upspinsos",
+		"github.com/u-root/u-root/cmds/vboot",
+		"github.com/u-root/u-root/cmds/wget",
+		"github.com/u-root/u-root/cmds/which",
+		"github.com/u-root/u-root/cmds/wifi",
+		"github.com/u-root/u-root/cmds/xinit",
+	}
+	for i := 0; i < len(pkgs); i++ {
+		for _, b := range blacklist {
+			if pkgs[i] == b {
+				pkgs = append(pkgs[:i], pkgs[i+1:]...)
+			}
+		}
+	}
+
+	return pkgs
+}
+
+// TestGoTest effectively runs "go test ./..." inside a QEMU instance. The
+// tests run as root and can do all sorts of things not possible otherwise.
+func TestGoTest(t *testing.T) {
+	// Create a temporary directory.
+	tmpDir, err := ioutil.TempDir("", "uroot-integration")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Statically build tests and add them to the temporary directory.
+	pkgs := testPkgs(t)
+	os.Setenv("CGO_ENABLED", "0")
+	testDir := filepath.Join(tmpDir, "tests")
+	for _, pkg := range pkgs {
+		testFile := filepath.Join(testDir, path.Base(pkg))
+		cmd := exec.Command("go", "test", "-ldflags", "-s -w", "-c", pkg, "-o", testFile)
+		cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
+		if err := cmd.Run(); err != nil {
+			t.Fatalf("could not build %s: %v", pkg, err)
+		}
+	}
+
+	// Create the CPIO and start QEMU.
+	tmpDir, q := testWithQEMU(t, options{
+		uinitName: "gotest",
+		tmpDir:    tmpDir,
+	})
+	defer cleanup(t, tmpDir, q)
+
+	// Check that each test passed.
+	bases := []string{}
+	for _, pkg := range pkgs {
+		bases = append(bases, path.Base(pkg))
+	}
+	sort.Strings(bases) // Tests are run and checked in sorted order.
+	t.Log("TAP: TAP version 12")
+	t.Logf("TAP: 1..%d", len(bases))
+	for i, base := range bases {
+		runMsg := fmt.Sprintf("TAP: # running %d - %s", i, base)
+		passMsg := fmt.Sprintf("TAP: ok %d - %s", i, base)
+		failMsg := fmt.Sprintf("TAP: not ok %d - %s", i, base)
+
+		t.Log(runMsg)
+		if err := q.ExpectTimeout(passMsg, 2*time.Second); err == nil {
```

Comment:
Can you try this with a purposely failing test? Won't a test failure result in the Expect timing out, which may cause the next expect to miss the output of the next test? I think you may have a race here

Reply:
If I remember, the semantics of expect is to keep the output in a buffer until there is a match (when an expect fails due to a timeout, it does not affect the buffer). The TAP strings contain ids, so the next test won't match it by mistake.

Regardless, I can make this a regex which matches both "not ok" and "ok", then check which one it is. This would also remove the 2 second wait when the test fails.

Reply:
Oh also, what would happen if a test took 10 seconds to run? It would cause the next 4 tests to fail with it. Maybe this is what you meant.

To fix this, I can add a timeout for each test in uinit. If the test takes longer than 2 seconds, the process gets killed.

Reply:
Your concern is good.

Reply:
I just pushed something which should make it more robust.